### PR TITLE
do not inline `grow_amortized`

### LINKED ISSUE #78471: Code bloat from RawVec::grow_amortized
The change in #72227 made non-PGO builds of Firefox's libxul.so's .text section grow by close to 1MB. libxul.so has a ~89MB section, which makes it a > 1% size increase in a library where about only 16% of the code is rust, which makes it a > 7% (!) increase in size. This comes from the fact that essentially all uses of Vec::push inline this all for some reason, and it's used all across the place.

### PR DESCRIPTION
**Pull request by lcnr**
should fix #78471, let's see the impact on perf here though.

### CODE REVIEW

File: library/alloc/src/raw_vec.rs [hunk 1]
```diff
@@ -389,6 +389,9 @@ impl<T, A: AllocRef> RawVec<T, A> {
     // so that all of the code that depends on `T` is within it, while as much
     // of the code that doesn't depend on `T` as possible is in functions that
     // are non-generic over `T`.
+    //
+    // We never inline this function to keep `Vec::push` small.
+    #[inline(never)]
```
Comment (m-ou-se) [inline library/alloc/src/raw_vec.rs:394]: In many places, `#[cold]` is used for this, sometimes together with `#[inline(never)]` on the same function. Is there some guidance somewhere on which to use where?

If these kind of functions should have `#[inline(never)]` instead (or in addition to) of `#[cold]`, some of them (e.g. `SyncOnceCell::initialize`) should probably be updated.

Comment (lcnr) [reply library/alloc/src/raw_vec.rs:394]: hmm, I am not really sure. It probably depends on where it is used and we would need an additional perf run :laughing: 

I do feel like this method is kind of special as it is extremely frequently used. Not sure if `SyncOnceCell` is used frequently enough for this to be noticeable. Could be worth trying though

**Comment by lcnr** [comment]
looking at the perf cost this does not seem desirable to me

**Comment by glandium** [comment]
Can you try never inlining finish_grow? because it appears to always be inlined.

**Comment by lcnr** [comment]
won't do so myself, but feel free to open a PR doing this and assign me using `r? @lcnr` and we can then look at the perf results for that

**Pull request closed without merging**